File: python/app/services/auth_service.py
"""Auth flows: register (customer role only — self-signup role is always
`customer`, matching the reference app's hidden-role register page; a
`seller` account is assumed pre-provisioned, see README), login, logout.
"""

import asyncio
from typing import Any

from app.mudbase_client import login_sync, logout_sync, register_with_role_sync


async def register_customer(
    email: str,
    password: str,
    first_name: str,
    last_name: str,
    agreed_to_terms: bool,
) -> dict[str, Any]:
    return await asyncio.to_thread(
        register_with_role_sync,
        "customer",
        email,
        password,
        first_name,
        last_name,
        agreed_to_terms,
    )


async def login(email: str, password: str) -> dict[str, Any]:
    return await asyncio.to_thread(login_sync, email, password)


async def logout(access_token: str) -> None:
    await asyncio.to_thread(logout_sync, access_token)

File: java/src/main/java/dev/mudbase/showcase/ecommerce/service/AuthService.java
package dev.mudbase.showcase.ecommerce.service;

import dev.mudbase.showcase.ecommerce.auth.AuthSession;
import dev.mudbase.showcase.ecommerce.auth.SessionAuthService;
import dev.mudbase.showcase.ecommerce.mudbase.AuthResult;
import dev.mudbase.showcase.ecommerce.mudbase.MudbaseAuthClient;
import jakarta.servlet.http.HttpSession;
import org.springframework.stereotype.Service;

/**
 * Registration is always role-hidden and customer-only from the public register form, matching
 * the reference app; a `seller` account is provisioned out-of-band and simply signs in through
 * the same login form (login isn't role-scoped).
 */
@Service
public class AuthService {

  private final MudbaseAuthClient authClient;
  private final SessionAuthService sessionAuthService;
  private final CartService cartService;

  public AuthService(MudbaseAuthClient authClient, SessionAuthService sessionAuthService, CartService cartService) {
    this.authClient = authClient;
    this.sessionAuthService = sessionAuthService;
    this.cartService = cartService;
  }

  public AuthSession registerCustomer(
      HttpSession session, String email, String password, String firstName, String lastName, boolean agreedToTerms) {
    AuthResult result = authClient.registerWithRole("customer", email, password, firstName, lastName, agreedToTerms);
    return establishAndMigrateCart(session, result);
  }

  public AuthSession login(HttpSession session, String email, String password) {
    AuthResult result = authClient.login(email, password);
    return establishAndMigrateCart(session, result);
  }

  private AuthSession establishAndMigrateCart(HttpSession session, AuthResult result) {
    sessionAuthService.establish(session, result);
    AuthSession established = sessionAuthService.current(session).orElseThrow();
    if (established.isCustomer()) {
      cartService.migrateGuestCartToServer(session, established);
    }
    return established;
  }

  public void logout(HttpSession session) {
    sessionAuthService.logout(session);
  }
}
